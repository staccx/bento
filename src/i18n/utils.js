import SanityBlockContent from "../sanity/components/SanityBlockContent"
import Text from "../components/Text/Text/Text"
import React from "react"

export const isBlock = item =>
  item &&
  item.hasOwnProperty("_type") &&
  (item._type === "block" || item._type === "image")

export const isRichText = val =>
  Array.isArray(val) &&
  val.some(
    arr =>
      typeof arr !== "string" &&
      Array.isArray(arr) &&
      arr.some(item => item._type)
  )

export const getComponent = val => {
  return Array.isArray(val) ? (
    val.map(getComponent)
  ) : isBlock(val) ? (
    <SanityBlockContent key={val._key} blocks={val} />
  ) : (
    <Text key={val}>{val}</Text>
  )
}

export const handleArray = (value, data, children, translate) => {
  const values = value.map(
    (k, index) => translate(k, data) || React.Children(children)[index]
  )

  if (typeof children === "function") {
    return children(values)
  }

  return values.map(getComponent)
}
