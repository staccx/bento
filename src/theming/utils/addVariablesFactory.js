import set from "lodash.set"
import { Theme, themes } from "../index"
import { themeLogger } from "../theme.logger"
import { getDisplayName } from "./getDisplayName"

export const addVariablesFactory = component => (
  variables,
  themeName = "_default"
) => {
  if (!themes.hasOwnProperty(themeName)) {
    themeLogger.warn(
      `Theme with name ${themeName} does not exists. One will be created`
    )
    themes[themeName] = new Theme()
  }
  const theme = themes[themeName]

  if (!component) {
    throw new Error("No component")
  }

  if (!variables) {
    throw new Error("addVariables called with no variables")
  }

  for (const key of Object.keys(variables)) {
    const variant = variables[key]

    const styles = {}
    set(styles, `variables.${getDisplayName(component)}.${key}`, variant)
    theme.append(styles)
  }
}
