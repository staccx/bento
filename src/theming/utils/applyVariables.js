import get from "lodash.get"
import { VARIANT_DEFAULT } from "../themeContants"
import { themeLogger } from "../theme.logger"

export const applyVariables = name => props => {
  const { variant = VARIANT_DEFAULT, theme } = props
  if (!theme) {
    themeLogger.warn("No theme!", variant)
    return null
  }

  if (!name) {
    themeLogger.warn("No name!", variant, theme)
    return null
  }

  const styles = get(theme, `variables.${name}`)

  if (!styles || !styles[variant]) {
    themeLogger.warn(
      "Could not find any styles for variant",
      variant,
      `variables_${name}`
    )
    return null
  }

  return styles[variant]
}
