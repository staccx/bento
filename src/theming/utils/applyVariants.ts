import deepfind from "./deepfind"
import { VARIANT_DEFAULT } from "../themeContants"

/**
 * Apply variants for a component. Allows for multiple variants
 * @param themeProps
 * @returns {function} that styled-components uses in the resolving.
 */
export default themeProps => props => {
  const { variant = VARIANT_DEFAULT, theme } = props

  if (!theme) {
    console.warn("No theme!", variant)
    return null
  }

  if (!themeProps) {
    console.error("No theme props!", theme, variant, props)
    return null
  }

  const name = themeProps.hasOwnProperty("name") ? themeProps.name : themeProps

  if (!name) {
    console.warn("No name!", variant, theme)
    return null
  }

  return getVariants(theme, variant, name)
}

/**
 * Get variants.
 * @example: resolveVariants(theme, "_default", "button", themeProps)
 * @param theme
 * @param variant - string, array, function or object
 * @param name - name of the style group e.g "button"
 * @returns {object, array}
 */
const getVariants = (theme, variant, name) => {
  const type = typeof variant
  if (Array.isArray(variant)) {
    console.info("Variant is an array", variant)
    return variant.map(v => getVariant(theme, v, name))
  } else if (type === "function") {
    console.info("Variant is a function", variant)
    return getVariant(theme, variant({ name, theme }), name)
  } else if (type === "string") {
    console.info("Variant is a string", variant)
    return getVariant(theme, variant, name)
  } else if (type === "object") {
    console.info("Variant is an object", variant)
    const keys = Object.keys(variant)
    return keys.map(key => getVariant(theme, variant[key], key))
  }
  console.warn("Unsupported variant type", type, variant, name)
  return null
}
/**
 * Get single variant.
 * @param theme
 * @param variant
 * @param name
 * @returns {object}
 */
const getVariant = (theme, variant, name) => {
  const styles = deepfind(theme, name)
  if (!styles || !styles[variant]) {
    console.warn("Could not find any styles for variant", variant, name)
    return null
  }

  return styles[variant]
}
