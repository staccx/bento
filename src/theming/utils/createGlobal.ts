export default variants => {
  return {
    global: {
      ...variants
    }
  }
}
