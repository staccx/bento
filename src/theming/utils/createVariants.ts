export default (variants, style) => {
  const name = style && style.hasOwnProperty("name") ? style.name : style
  return {
    [name]: {
      ...variants
    }
  }
}
