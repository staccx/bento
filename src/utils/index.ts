export { stringIncludes } from "./stringIncludes"
export { distinct } from "./distinct"
export { flatten } from "./flatten"
export { partition } from "./partition/partition"
