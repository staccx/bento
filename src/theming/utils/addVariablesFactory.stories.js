import {addVariablesFactory} from "./addVariablesFactory";
import { RadioPill } from "../../components";
import { css } from "styled-components";

export default {
  title: "Utils/Add Variables Factory",
  component: addVariablesFactory,
  parameters: {
    docs: {
      description: {
        component: `
        # Add Variables.
          Inject css variables into the theme
        `
      },
    }
  },
}

export const NormalUsage = () => {
  const variables = RadioPill.addVariables({
    variant1: css`
      --background-color: 2px;
    `
  })



  return null


}
