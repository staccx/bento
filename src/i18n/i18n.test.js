import React from "react"
import { render, cleanup, waitForElement, act } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect"
import renderer from "react-test-renderer"
import { isBlock } from "./utils"
import {
  // withI18n,
  I18nConsumer,
  // SanityBackend,
  // useI18n,
  Provider
  // Translate,
  // TransformFG
} from "./index"

afterEach(cleanup)

const data = {
  _createdAt: "2019-06-19T10:19:01Z",
  _id: "37fd9bcc-100b-42f8-9731-fb9d9675069f",
  _rev: "3IttJ7VKu6Yi5HHVPkhXAY",
  _type: "translations",
  _updatedAt: "2019-06-19T10:19:01Z",
  i18nKey: {
    _type: "slug",
    current: "rik-tekst"
  },
  name: "Rik tekst",
  value: [
    {
      _key: "9c832c54c0fb",
      _type: "localeRichText",
      nb: [
        {
          _key: "42b883c30ffa",
          _type: "block",
          children: [
            {
              _key: "42b883c30ffa0",
              _type: "span",
              marks: [],
              text: "Dette er "
            },
            {
              _key: "42b883c30ffa1",
              _type: "span",
              marks: ["em"],
              text: "rik"
            },
            {
              _key: "42b883c30ffa2",
              _type: "span",
              marks: [],
              text: " "
            },
            {
              _key: "42b883c30ffa3",
              _type: "span",
              marks: ["strong"],
              text: "tekst"
            }
          ],
          markDefs: [],
          style: "normal"
        }
      ]
    }
  ]
}

describe("i18n", () => {
  describe("utils", () => {
    it("isBlock", () => {
      expect(isBlock("test")).toBe(false)
      expect(isBlock(data.value[0].nb[0])).toBe(true)
    })
  })
  describe("Context", () => {
    describe("Provider", () => {
      it("Consumer shows value from provider", () => {
        let tree = null
        act(() => {
          tree = render(
            <Provider language="en">
              <I18nConsumer>
                {value => {
                  console.log(value.ready)
                  return <span>Ready: {value.ready}</span>
                }}
              </I18nConsumer>
            </Provider>
          )
          const { getByText } = tree
          waitForElement(() => getByText(/^Ready:/)).then(wait => {
            expect(wait.textContent).toBe("Ready: true")
          })
        })
      })
    })
  })
  describe("Snapshots", () => {
    it("Normal render", () => {
      let tree = null
      renderer.act(() => {
        tree = renderer.create(
          <Provider language="en">
            <I18nConsumer>
              {value => {
                return <span>Ready: {value.ready}</span>
              }}
            </I18nConsumer>
          </Provider>
        )
      })
      expect(tree.toJSON()).toMatchSnapshot()
    })
  })
})
