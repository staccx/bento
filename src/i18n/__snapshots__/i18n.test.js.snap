// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`i18n Snapshots Normal render 1`] = `null`;
